CorsixTH/Lua/dialogs/build_room.lua

```diff
@@ -102,16 +102,18 @@ function UIBuildRoom:UIBuildRoom(ui)
   self:makeTooltip(_S.tooltip.build_room_window.cost, 160, 228, 282, 242)
 end
 
+--! Checks what rooms are now available to build
 function UIBuildRoom:updateBuildableRooms()
   local app = self.ui.app
   self.category_rooms = {}
   for i, category in ipairs({"diagnosis", "treatment", "clinics", "facilities"}) do
     local rooms = {}
     self.category_rooms[i] = rooms
-    for _, room in ipairs(app.world.available_rooms) do
+    for j, room_idx in ipairs(self.ui.hospital.room_discovery) do
+      local room = room_idx.room
       -- NB: Unimplemented rooms are hidden unless in debug mode
       if (app.config.debug or room.class) and room.categories[category] and
-          self.ui.hospital.discovered_rooms[room] then
+          self.ui.hospital.room_discovery[j].is_discovered then
         rooms[#rooms + 1] = room
       end
     end
```

Comment:
I think the hospital should do this rather than the UI.

One thing to note here is that you don't introduce a shortcut variable to the long `self.ui.hospital.room_discovery` sequence. If you find repetitive use of the same thing, it's often useful to introduce a shorter notation for it.